tools/src/test/scala/caliban/tools/SchemaWriterSpec.scala

```diff
@@ -630,6 +627,75 @@ object SchemaWriterSpec extends ZIOSpecDefault {
         |  ) derives caliban.schema.Schema.SemiAuto
         |
         |}""".stripMargin
+    ),
+    (
+      "inherits field with args",
+      gen(
+        """
+        |interface Character {
+        |    friendsConnection(first: Int, after: ID): FriendsConnection!
+        |}
+        |type Human implements Character {
+        |    friendsConnection(first: Int, after: ID): FriendsConnection!
+        |}
+        |type Droid implements Character {
+        |    friendsConnection(first: Int, after: ID): FriendsConnection!
+        |}""",
+        addDerives = true
+      ),
+      """import caliban.schema.Annotations._
+        |
+        |object Types {
+        |  final case class CharacterFriendsConnectionArgs(first: scala.Option[Int], after: scala.Option[ID])
+        |      derives caliban.schema.Schema.SemiAuto,
+        |        caliban.schema.ArgBuilder
+        |  final case class Human(friendsConnection: CharacterFriendsConnectionArgs => FriendsConnection) extends Character
+        |      derives caliban.schema.Schema.SemiAuto
+        |  final case class Droid(friendsConnection: CharacterFriendsConnectionArgs => FriendsConnection) extends Character
+        |      derives caliban.schema.Schema.SemiAuto
+        |
+        |  @GQLInterface
+        |  sealed trait Character extends scala.Product with scala.Serializable derives caliban.schema.Schema.SemiAuto {
+        |    def friendsConnection: CharacterFriendsConnectionArgs => FriendsConnection
+        |  }
+        |
+        |}""".stripMargin
+    ),
+    (
+      "EffectStrategy.EffectfulIfTaggedWith",
+      gen(
+        """
+        |directive @lazy on FIELD_DEFINITION
+        |
+        |type Foo {
+        |  bar: String!
+        |  baz: String! @lazy
+        |}""",
+        effectStrategy = EffectStrategy.EffectfulIfTaggedWith("lazy")
+      ),
```

Comment:
I'm thinking whether we should consider all fields marked with `@lazy` as effectful, regardless of the effectStrategy. I can't really imagine any other usage of that directive that would mean something different

Reply:
I didn't want to standardize on a name in this first version.

So to be clear: In order for `EffectStrategy.EffectfulIfTaggedWith` have an effect currently:
- you need to set it up with a name, `lazy` in this case
- define the `lazy` annotation yourself
- tag a field with it

An alternative approach would be to standardize on one name, and not make it configurable

Reply:
Maybe just choosing a name is simpler? And it has the advantage of working out of the box with the sbt plugin (exposing options to the plug-in is rather painful).

Reply:
I think let's use either `lazy` or `lazyField`, and _if_ users want to be able to customise it for some reason we can add the customisation back.

On that note, what do you think about adding an input argument to that directive with a default argument `(effectful: Boolean = true)`, and generate a `() => A` when that is set to `false`?

Reply:
Why not use `effectType: Option[String]` which lets you specify the effect type. That way you can mix different ones like `Task`/`UIO` or `ZIO`/`ZQuery`. And `None` could be for `() =>`

Reply:
Yeah that sounds like a better idea

Reply:
Some followups before I get back to work on this:

> Maybe just choosing a name is simpler? And it has the advantage of working out of the box with the sbt plugin (exposing options to the plug-in is rather painful). 

First complication: we only want to generate effect types for *server* code, it does not make sense on the client. So some parameterization is needed for the codegen either way. I can inline the `EffectStrategy.EffectfulIfTaggedWith` strategy, but how would you want the client/server distinction to be made?

> I think let's use either lazy or lazyField, and if users want to be able to customise it for some reason we can add the customisation back.

I think `lazy` is a good name for scala developers, but I don't think it communicates well to other people who consume the schema with this directive thrown in. I also don't have a better alternative at the top of my head, which is why I wanted to delay that decision and let it be user-configurable. Other suggestions?

> Why not use effectType: Option[String] which lets you specify the effect type. That way you can mix different ones like Task/UIO or ZIO/ZQuery, which is more powerful than the current way. And None could be for () =>

Do you mean to expose `effectType` in the _directive_? It seems to me like oversharing a lot in the public API.

I would want this to bake a bit honestly before we commit to it.
- we'll need the directive to apply to top-level queries as well. Those are unconditionally effectful currently, so the mechanism may be a bit different.
- need a fallback if effect type is not specified
- I think there will be some complexity around mixing effect types. `F` needs to be surrounded by a `class`, `IO` needs non-global type-class evidence in order to output `derives Schema`, `ZQuery` has multiple type arguments.

Reply:
I'm fine with just adding a simple directive for now and see if people express the need to customize the effect type later. Also agree it makes sense to make it configurable.

Reply:
> Do you mean to expose effectType in the directive? It seems to me like oversharing a lot in the public API.

Since these directives are used only for codegen customization, do you think we could (should?) exclude them from introspection?

Having said that, I agree with @ghostdogpr, let's start with something simple and we can add more customization in the future

Reply:
ok, I simplified the PR now. deleted `EffectStrategy` and made the codegen look for a `lazy` annotation. I realized the client sourcegen was a completely separate thing now :) 

WDYT?